### PR TITLE
feat: logging control operation

### LINKED ISSUE #179: feat: [Trainer controller] Add custom logging operation
## Is your feature request related to a problem? Please describe.

There is no operation to introduce custom logging formats.

## Describe the solution you'd like

Custom logging operation allows user to introduce new log lines on specific trigger events.

## Describe alternatives you've considered

NA

## Additional context

NA

### PR DESCRIPTION
**Pull request by seshapad**
### Description of the change

Logging control operation to inject logs into the training loop.

### Related issue number

"Closes #179"

### How to verify the PR

Use the following trainer controller configuration to try out the logging control:
```
controller_metrics:
  - name: state
    class: TrainingState
operations:
  - name: logcontrolstep
    class: LogControl
    arguments:
      log_format: 'This is a test log format [{event_name}] => {tc_metrics[state]}'
      log_level: warning
controllers:
  - name: log-controller-step
    triggers:
      - on_step_end
    rule: 'True'
    operations:
      - logcontrolstep.should_log
```

### Was the PR tested

- [x] I have added >=1 unit test(s) for every new method I have added.
- [x] I have ensured all unit tests pass

### CODE REVIEW

File: tuning/trainercontroller/operations/logcontrol.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+logger = logging.get_logger(__name__)
+logger.setLevel(level=logging.DEBUG)
+
+VALID_LOG_LEVELS = {
+    "ERROR": logging.ERROR,
+    "WARNING": logging.WARNING,
+    "INFO": logging.INFO,
+    "DEBUG": logging.DEBUG,
+}
```
Comment (ashokponkumar) [inline tuning/trainercontroller/operations/logcontrol.py:16]: Why are we redefining the levels here instead of using it from the logging library?

File: tuning/trainercontroller/operations/logcontrol.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+        Args:
+            control: TrainerControl. Data class for controls.
+            kwargs: List of arguments (key, value)-pairs
+        """
+        log_msg = self.log_format.format(
+            event_name=event_name,
+            tc_metrics=tc_metrics,
+            args=args,
+            **kwargs,
```
Comment (ashokponkumar) [inline tuning/trainercontroller/operations/logcontrol.py:56]: should we flatten this to make rule and format to look similar?

**Review by ashokponkumar — CHANGES_REQUESTED** [review]

File: tuning/trainercontroller/operations/logcontrol.py (hunk 1 above)
Comment (seshapad) [reply tuning/trainercontroller/operations/logcontrol.py]: Addressed.

File: tuning/trainercontroller/operations/logcontrol.py (hunk 2 above)
Comment (seshapad) [reply tuning/trainercontroller/operations/logcontrol.py:49]: Could you please elaborate on this?

Comment (seshapad) [reply tuning/trainercontroller/operations/logcontrol.py]: @ashokponkumar I have addressed the above comment. Following are the keywords we need to avoid using as trainer controller metric names as they conflict when passed with `**kwargs`:
```
model 
tokenizer
optimizer
lr_scheduler
train_dataloader
eval_dataloader 
metrics
logs
```
Source is [HF Callback class definition](https://github.com/huggingface/transformers/blob/868d36d29ec132deeaaf8571b25b6a1b911d0145/src/transformers/trainer_callback.py#L261).

File: examples/trainercontroller_configs/log_controller.yaml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+controller_metrics:
+  - name: state
+    class: TrainingState
+operations:
+  - name: logcontrolstep
+    class: LogControl
+    arguments:
+      log_format: 'This is a test log format [{event_name}] => {tc_metrics[state]}'
```
Comment (ashokponkumar) [inline examples/trainercontroller_configs/log_controller.yaml:8]: Shouldn't this be just `state`?
```suggestion
      log_format: 'This is a test log format [{event_name}] => {state}'
```

File: tuning/trainercontroller/controllermetrics/trainingstate.py [hunk 4]
```diff
@@ -49,7 +52,7 @@ def __init__(self, **kwargs):
                 "on_train_begin",
                 "on_evaluate",
             ],
-            **kwargs
+            **kwargs,
```
Comment (ashokponkumar) [inline tuning/trainercontroller/controllermetrics/trainingstate.py:55]: Is this the right convention?

**Review by ashokponkumar — CHANGES_REQUESTED** [review]
nit. Can you please fix them?

File: examples/trainercontroller_configs/log_controller.yaml (hunk 3 above)
Comment (seshapad) [reply examples/trainercontroller_configs/log_controller.yaml]: Addressed.

File: tuning/trainercontroller/controllermetrics/trainingstate.py (hunk 4 above)
Comment (seshapad) [reply tuning/trainercontroller/controllermetrics/trainingstate.py:55]: seems like it. Here is one [example](https://github.com/foundation-model-stack/fms-hf-tuning/blob/d7280073f8121ff2d0b160302092ddc13010852f/tuning/config/acceleration_configs/acceleration_framework_config.py#L78).

**Comment by ashokponkumar** [comment]
@kmehant PTAL

**Review by ashokponkumar — APPROVED** [review]

**Review by kmehant — APPROVED** [review]
Adds log control and a test.

**Comment by kmehant** [comment]
@seshapad lets rebase and we can merge it thanks
